### ISSUE TITLE
Give a concrete example

### ISSUE THREAD

**Issue by probonopd** [niess/linuxdeploy-plugin-python#5]
Thanks for making this, it looks very useful. Would you be able to give a concrete real-world example, e.g., on how to package https://github.com/probonopd/ptouch-770?

**Comment by niess** [comment]
@probonopd I'll try to do so. It probably wont work at first. But that should be fun :)

Actually, the primary focus for me was to pack a clean, isolated and ready to use Python work environment. E.g. for collaborative developments in Python using a same basis, etc. I was planning to use this in order to build a framework for the software of a Physics experiment.

**Comment by probonopd** [comment]
Cool @niess. That's how it starts - by scratching your itch. And then you realize that you have made something that is useful for a much larger community :+1:

**Comment by gmagno** [comment]
This is promising!
It would be great to see an example on how to bundle a python app with a few PyPI deps.
Cheers

**Comment by niess** [comment]
@gmagno In case you are still interested in this, there is now a new project: [python-appimage](https://github.com/niess/python-appimage) re-written in Python :P It fasten and simplifies the build of Python AppImages. For basic apps that package all their deps in wheels you can easily build an AppImage using the new project. Sample apps are provided in the `applications/` folder. You can build them as:

```
python-appimage build app applications/tasmotizer
```

Alternatively you can also download an existing Python AppImage from the [releases](https://github.com/niess/python-appimage/releases), extract it, pip install to it and rebuild the AppImage once done.

**Comment by gmagno** [comment]
@niess I love it! This is so useful and easy to use, thank you.

**Comment by niess** [comment]
@gmagno Thank you :)

I could add this AppImage as another example of application, and add it to the releases. Note that you can get the package from pip though. It should run with any Python starting from 2.7. Yet, on old systems using older Python versions it could help having this AppImage indeed.

It would be cool to have some utility to strip the images as well for specific applications. Maybe by removing packages that are not referenced by the main script?

**Comment by gmagno** [comment]
> I could add this AppImage as another example of application, and add it to the releases. Note that you can get the package from pip though. It should run with any Python starting from 2.7. Yet, on old systems using older Python versions it could help having this AppImage indeed.

Indeed. And since the user will be writing a python app, she will definitely have python, hence `pip install python-appimage` makes sense.
Generically speaking though, `python-appimage` can be seen as a deliverable: a CLI tool that people will use. I can see a few scenarios, e.g automation, where `create venv+activate venv+pip install+run` could be a nuisance compared to `wget+chmod+run`.

**Comment by niess** [comment]
This was my initial idea: that people would already have Python. Yet, In order to isolate `python-appimage` as CLI tool indeed it would make sense to have a specific AppImage. Installing with the `--user` scheme can be problematic with multiple versions of Python  and even more with various flavours of the same Python version -some of them running from AppImages. Isolating everything in venv is a bit tedious indeed.

I added a [specific issue](https://github.com/niess/python-appimage/issues/5) for this since I won't have time to do it right now.

**Comment by probonopd** [comment]
My use case:

I see some cool Python app somewhere on the web. It talks about `pip install`.

Now I want to make an AppImage of that application, ideally in the most automated way possible. Something like

`./pipToAppImage_1.0_x86_64.AppImage nameofthepip`

**Comment by niess** [comment]
@probonopd  I think that this is almost the case with `python-appimage` :P If the app can be pip installed then it should run from the AppImage as well, provided that the Python package conforms to one of the manylinux specs. One just needs to extract the image, pip install with the AppRun, modify the AppRun in order to start the Python app and package it back. The `build app` mode of `python-appimage` basically does this.

The most tedious part actually is updating the metadata, i.e. the desktop file etc. That's why I finally came up with this solution of having a Python application directory, e.g. as [here](https://github.com/niess/python-appimage/tree/master/applications/tasmotizer). As is done with `appimagetool`.

I could add a mode doing as you said, i.e. basically allowing to pip install to an existing Python AppImage transparently. That would be very easy to use indeed and easy to implement. But then people wouldn't update the metadata. Would they? Also when you `pip install` a package you actually don't get the information of what Python script the AppImage should run. If the package only installs a single script then it should be manageable. But if it installs several scripts how to know which one to run?

**Comment by probonopd** [comment]
```
./python3.8.2-cp38-cp38-manylinux1_x86_64.AppImage -m pip install tasmotizer
./python3.8.2-cp38-cp38-manylinux1_x86_64.AppImage -m tasmotizer
```

can be used to just get and run an application.

To turn it into an AppImage more information for the metadata is needed, this is correct. Auto-populating desktop files and such is not the right thing to do, as lazy developers would leave the defaults there like you say.

So the current implementation is probably the best approach.
